pkg/secrets/config/config_test.go

```diff
@@ -1,76 +1,395 @@
+	},
+	{
+		description: "if RequiredK8sSecrets is not configured in K8s Secrets mode, an error is returned",
+		envAndAnnots: map[string]string{
+			"MY_POD_NAMESPACE":               "test-namespace",
+			"conjur.org/secrets-destination": "k8s_secrets",
+		},
+		assert: assertErrorInList(errors.New(messages.CSPFK048E)),
+	},
+	{
+		description: "if envVar 'SECRETS_DESTINATION' is malformed in the absence of annotation 'conjur.org/secrets-destination', an error is returned",
+		envAndAnnots: map[string]string{
+			"MY_POD_NAMESPACE":    "test-namespace",
+			"SECRETS_DESTINATION": "invalid",
+		},
+		assert: assertErrorInList(fmt.Errorf(messages.CSPFK005E, "SECRETS_DESTINATION")),
+	},
+}
+
+type newConfigTestCase struct {
+	description string
+	settings    map[string]string
+	assert      func(t *testing.T, config *Config)
+}
+
+var newConfigTestCases = []newConfigTestCase{
+	{
+		description: "a valid map of annotation-based Secrets Provider settings returns a valid Config",
+		settings: map[string]string{
+			"MY_POD_NAMESPACE":               "test-namespace",
+			"conjur.org/secrets-destination": "k8s_secrets",
+			"conjur.org/k8s-secrets":         "- secret-1\n- secret-2\n- secret-3\n",
+			"conjur.org/retry-count-limit":   "10",
+			"conjur.org/retry-interval-sec":  "20",
+		},
+		assert: assertGoodConfig(&Config{
+			PodNamespace:       "test-namespace",
+			StoreType:          "k8s_secrets",
+			RequiredK8sSecrets: []string{"secret-1", "secret-2", "secret-3"},
+			RetryCountLimit:    10,
+			RetryIntervalSec:   20,
+		}),
+	},
+	{
+		description: "a valid map of envVar-based Secrets Provider settings returns a valid Config",
+		settings: map[string]string{
+			"MY_POD_NAMESPACE":    "test-namespace",
+			"SECRETS_DESTINATION": "k8s_secrets",
+			"K8S_SECRETS":         "secret-1,secret-2, secret-3",
+			"RETRY_COUNT_LIMIT":   "10",
+			"RETRY_INTERVAL_SEC":  "20",
+		},
+		assert: assertGoodConfig(&Config{
+			PodNamespace:       "test-namespace",
+			StoreType:          "k8s_secrets",
+			RequiredK8sSecrets: []string{"secret-1", "secret-2", "secret-3"},
+			RetryCountLimit:    10,
+			RetryIntervalSec:   20,
+		}),
+	},
+	{
+		description: "settings configured with both annotations and envVars defer to the annotation value",
+		settings: map[string]string{
+			"MY_POD_NAMESPACE":               "test-namespace",
+			"SECRETS_DESTINATION":            "k8s_secrets",
+			"conjur.org/secrets-destination": "file",
+			"K8S_SECRETS":                    "secret-1,secret-2,secret-3",
+		},
+		assert: assertGoodConfig(&Config{
+			PodNamespace:       "test-namespace",
+			StoreType:          "file",
+			RequiredK8sSecrets: []string{},
+			RetryCountLimit:    DEFAULT_RETRY_COUNT_LIMIT,
+			RetryIntervalSec:   DEFAULT_RETRY_INTERVAL_SEC,
+		}),
+	},
+}
+
+func TestValidateAnnotations(t *testing.T) {
+	for _, tc := range validateAnnotationsTestCases {
```

Comment:
This is COMPLETELY a personal preference sort of thing, but in the past, I've created table-driven tests where all of these are defined in one test function:

- The test case struct (this struct can be anonymous, if it's included in an instantiation of this struct, i.e. the test list)
- The test case list (e.g. `validateAnnotationsTestCases`)
- The iteration/running through all of the test cases

The benefit to this is that all definitions required for the test are gathered into one spot (one function). The downside is that the test function ends up with a LOT of lines (less visual breaks).